### DISCUSSION TITLE
what is IsMarked?

### DISCUSSION THREAD

**Discussion by RetroGaming12221** [tui-cs/Terminal.Gui#2321] [Q&A] [answered]
help me, idk wth is "IsMarked" function is

**Comment by tznind** [comment, marked as answer]
It indicates whether user has ticked the item when `AllowsMarking` is true on a `ListView`.  If you are writing your own custom `IListDataSource` and you are not ever using any instances that have ticking enabled then you can ignore it.

![markgif](https://user-images.githubusercontent.com/31306100/216986702-18c961ce-edec-4fc2-9d39-8925e50588ad.gif)
_Space toggles IsMarked while Enter completes selection (see code below)_

```csharp
using Terminal.Gui;

Application.Init();

var w = new Window();
var lv = new ListView(new List<string> { "fish", "cat" })
{
    Width = 10,
    Height = 5,
    AllowsMarking = true
};

w.Add(lv);

lv.KeyPress += k =>
{
    if (k.KeyEvent.Key == Key.Enter)
    {
        var chosen = lv.Source.ToList().Cast<string>().Where((o, idx) => lv.Source.IsMarked(idx)).ToArray();
        MessageBox.Query("Chosen", "You chose " + string.Join(",", chosen), "Ok");
        k.Handled = true;
    }
};


Application.Run(w);
Application.Shutdown();

```

**Reply by RetroGaming12221** [reply]
oh wait, u can use a normal list?

**Reply by tznind** [reply]
Yes `ListView` has a constructor that takes a regular `IList` (which includes the generic).

Also, if you want to change the list after construction, then you can use `SetSource` method to give it a different list:
```csharp
lv.SetSource(new List<string> { "cat", "dog", "bazooka" });
```

You only need to implement `IListDataSource` if you are wanting to do fancy rendering (or otherwise want to change standard behaviours).  

Only thing to take care of is that you do not change the list during rendering.  You only have to worry about that if you have an async process adding to the list.  Easiest way to avoid it is to just pass a copy of the list to the `ListView`.

**Reply by RetroGaming12221** [reply]
oh lol, thank you for helping me. The documentation really confused me even after reading it a few times

**Reply by tznind** [reply]
> oh lol, thank you for helping me. The documentation really confused me even after reading it a few times

No worries.  The docs are quite technical for ListView.  By the way there is an open source designer https://github.com/gui-cs/TerminalGuiDesigner which might be useful if you have a lot of popups or UIs to create.  It includes support for setting up basic lists.  For example:

![designerlists](https://user-images.githubusercontent.com/31306100/217004732-f723e2bb-29d5-4dbf-a289-06c040bcfa1c.gif)

**Reply by RetroGaming12221** [reply]
:O. Thank you so much!
